articles/events_executor.md

```diff
@@ -0,0 +1,233 @@
+
+Whenever a node or a callback group are removed from the executor using the corresponding APIs (`remove_node()` and `remove_callback_group()`), the `EventsExecutor` needs to clean up the associated entities to make them stop sending events.
+Entities may be eventually added back again to the same executor or to a different one and they need to be able to start again producing events.
+This requires that the initialization procedure that is performed when entities are added to an executor is completely reversible.
+
+The events-based approach requires particular care in handling the case where entities are destroyed while still associated with an executor.
+An entity may push an event into the `EventsQueue` and then get immediately destroyed.
+This may happen before the `EventsExecutor` start to process those events or while it is processing them.
+
+The current implementation addresses this problem by providing an additional function pointer to each associated entity.
+This function pointer will be called whenever the entity's destructor is invoked and it will result in clearing the `EventsQueue` from any event that was generated by this entity. Note that this may block until the `EventsQueue` is done processing events.
+Alternative implementations may involve the `EventsExecutor` to keep weak pointers to the entities and lock them before executing any event.
+
+### Open Issues
+
+The `EventsQueue` does not have a knowledge of QoS settings, this may cause some issues in some particular situations.
+
+##### Unbound growth of event queue
+
+As long as events are pushed into the queue, this will keep growing, regardless of the actual history size of the underlying middleware entity.
+This may cause several events to accumulate, for example while the `EventsExecutor` is not spinning.
+
+
+##### Invalid ordering corner case
+
+The current design may fail to provide a correct ordering of events in some corner case situations.
+In particular, if an entity pushes a number of events greater than its QoS history size while the `EventsExecutor` is busy processing events, then the ordering may be compromised.
+
+Consider the following example of a system with two subscriptions A and B. Subscription A has a QoS history size of 1.
+While the `EventsExecutor` is busy processing events, the following events accumulates into the `EventsQueue`:
+ - Event 1 from Subscription A
+ - Event 2 from Subscription B
+ - Event 3 from Subscription A
+ 
+When the `EventsExecutor` is ready to process these new events, it will start from the first pushed, i.e. the event 1 from Subscription A.
+However, when calling the `execute_subscription()` API, the message that generated event 1 will not be available anymore as it has been overwritten from the message that generated event 3.
+The `execute_subscription()` API will not fail, but rather it will process the message that generated event 3.
+This violates the ordering as event 2 should have been processed before that.
+
```

Comment:
The wait set and the event queue approach both indicate that there is work to do and the actual work is grabbed later. The event queue has "normally" already a temporal order and less "walk through" and sorting than the wait set, but is also not deterministic as you describe here. 

Determinism would be reached with a work queue or if the collected work is sorted before executed in case of  event queue and wait set. So determinism would be possible if you pay with some runtime. 

I'm wondering if it would be possible to have an intermediate API that provides a container with work to execute. If this is then optimized for performance, latency, determinism or however sorted is the job of the flexible part of the executor. But I'm not familiar enough with the upper layers of rcl and rclcpp to judge if this makes sense. This maybe goes in the direction of the out of tree proposal @ivanpauno had.

Reply:
Determinism would be reached with a work queue or if the collected work is sorted
    before executed in case of event queue and wait set. So determinism would be possible
    if you pay with some runtime.

Determinism could be reached implementing what @alsora proposed above:

    The events executor design would allow to attach IDs to the messages/events,
    but then the rclcpp layer should also provide an API execute_subscription(Subscription, id)
    that will execute only if it's the correct element.

Regarding

    I'm wondering if it would be possible to have an intermediate API that provides a container with work to execute.

The issue with a work queue is that a message taken to be processed later could be invalid by the time of processing. If the QoS depth of the queue is 1 (one) and you take the message, but another message arrives before processing it, the stored message in the work queue should not be processed.

Reply:
> Determinism could be reached implementing what @alsora proposed above:
> 
> ```
> The events executor design would allow to attach IDs to the messages/events,
> but then the rclcpp layer should also provide an API execute_subscription(Subscription, id)
> that will execute only if it's the correct element.
> ```

That's a pragmatic and efficient solution for a guaranteed temporal order. The middleware needs to provide a sneak preview for the id without taking the message but that's a solvable problem too  

> The issue with a work queue is that a message taken to be processed later could be invalid by the time of processing. If the QoS depth of the queue is 1 (one) and you take the message, but another message arrives before processing it, the stored message in the work queue should not be processed.

I see. That are the consequences of having two queues that are processed at different times. One in the middleware and one in the executor. The work queue looses the QoS depth, the event queue can fulfil it with the consequence of outdated NOP events as discussed above.   

What I meant with 

> ```
> I'm wondering if it would be possible to have an intermediate API that provides a container with work to execute.
> ```

is the question if it would make sense to have an executor top layer API that provides the work to do without the need to go back to the middleware again (and the races this entails). This allows the most determinism as the executor then would provide a plan that is guaranteed to be executed. But I fear this is too far away from where we are today

Reply:
> What I meant with ... is the question if it would make sense to have an executor top layer API that provides the work to do without the need to go back to the middleware again (and the races this entails).

We have considered that approach, as it is one step closer to get better performances too. But for the sake of simplicity for a first proposal, and taking into account that we are already getting great performance improvements, we decided to leave it as a follow up improvement for the future. Looking at the CPU profiling flamegraphs we did (using rclcpp IPC) it doesn't seem to be much overhead of "go back to the middleware again" as the message passing occurs only on RCLCPP layer. The approach could have a bigger impact when not using rclcpp IPC.